File: Problem2_PartitionArrayMaxSum.py
# Time Complexity: O(nk), where n - length of the array, k - maximum length of the sub-array
# Space Complexity: O(n)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

# Solution:

class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        if not arr or len(arr) == 0:
            return 0

        n = len(arr)
        # Array to store the maximum sum for every element
        dp = [0 for x in range(n)]

        # For every element, update the max sum of all subarrays from size 1 to k
        for i in range(n):
            maximum = arr[i]
            size = 1
            while size <= k and i - size + 1 >= 0:
                maximum = max(maximum, arr[i - size + 1])

                if i - size >= 0:
                    dp[i] = max(dp[i], dp[i - size] + maximum * size)
                else:
                    dp[i] = max(dp[i], maximum * size)

                size += 1

        # Last element gives the overall largest sum
        return dp[n - 1]

File: Problem2_PartitionArrayMaxSum.java
// Time Complexity : o(nk)
// Space Complexity : o(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach\

class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        
        if(arr==null||arr.length==0||k==0)
            return 0;
        
        int[] dp=new int[arr.length];
        
        dp[0]=arr[0];
        
        for(int i=1;i<arr.length;i++)
        {
           int max=arr[i];
            
            for(int j=1;j<=k && i-j+1>=0;j++)
            {
                max=Math.max(max,arr[i-j+1]); //incoming element
                
                if(i-j>=0)
                {
                    dp[i]=Math.max(dp[i],dp[i-j]+max*j);
                    
                }
                else
                {
                   dp[i]=Math.max(dp[i],max*j);
                }
            }
        }
        return dp[arr.length-1];
    }
}